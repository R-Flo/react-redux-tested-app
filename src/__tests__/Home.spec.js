import React from 'react'
import { shallow, mount, render } from 'enzyme';

import ConnectedHome,{Home} from '../components/Home';

import {Provider} from 'react-redux';
import '../setUpTests';

import {addInputs, subtractInputs} from '../actions/calculatorActions'
import {createStore} from 'redux'
import calculatorReducers from '../reducers/calculatorReducers'

describe('HOME snapshot',()=>{
    it('It renders without crashing', () => {
        const wrapper =  shallow(<Home output={10}/>);
        expect(wrapper).toMatchSnapshot();
    });

});

describe('HOME shallow description',()=>{
    let wrapper;
    const output = 10;

    //Permet avant chaque it de refaire un fresh shallow rendering
    // afin d'éviter les conflits avec les tests précédents
    beforeEach(()=>{
        wrapper = shallow(<Home output={output}/>)
    });

    it('Renders', () => {
       expect(wrapper.length).toEqual(1)
    });

    it('Contains header - h2', () => {
        expect(wrapper.contains(<h2>Using React and Redux</h2>)).toBe(true)
    });
    it('H2 header value ', () => {
        expect(wrapper.find('h2').get(0).props.children).toBe("Using React and Redux")
    });
    it('Contains input1', () => {
        expect(wrapper.find('input').at(0)
                .equals(<input type="text" placeholder="Input 1" ref="input1" />))
                .toBe(true)
    });
    it('Contains input2', () => {
        expect(wrapper.find('input').at(1)
                .equals(<input type="text" placeholder="Input 2" ref="input2" />))
                .toBe(true)
    });
     it('Contains output', () => {
        expect(wrapper.find('input[placeholder="Output"]').prop('value')).toEqual(output)
    });
    it('Contains button with id="add"', () => {
        expect(wrapper.find('button#add').type()).toEqual('button')
    });
    it('Contains button with id="subtract"', () => {
        expect(wrapper.find('button#subtract').type()).toEqual('button')
    });
});


describe('HOME connected to store',()=>{

    let store, wrapper;

    beforeEach(()=>{
        store = createStore(calculatorReducers);
        wrapper = shallow( <Provider store={store}><ConnectedHome /></Provider> ).dive();
    });

    it('Check store works', () => {
        store.dispatch(addInputs(500));
        expect(wrapper.find(Home).prop('output')).toBe(500)
    });

});

describe('HOME mounted',()=>{

    let store, wrapper;
    const fetchRandomNumber = jest.fn().mockResolvedValue(145);


    beforeEach(()=>{
        store = createStore(calculatorReducers);
        wrapper = mount( <ConnectedHome store={store} fetchRandomNumber={fetchRandomNumber}/>);
    });


    it('Calculate when Inputs are Filled and ADD is Clicked', () => {

        let input1 = wrapper.find('input').at(0);
        input1.instance().value = 20;

        let input2 = wrapper.find('input').at(1);
        input2.instance().value = 20;

        let addButton = wrapper.find('button').at(0);

        addButton.simulate('click');

        let output = wrapper.find('input').at(2);

        expect(output.prop('value')).toEqual(40);
    });


    it('Calculate when Inputs are Filled and ADD is Clicked', () => {
        let substractButton = wrapper.find('button').at(0);
    });

    it('fetch when asked', () => {

        let fetchButton = wrapper.find('button').at(2);

        fetchButton.simulate('click');

        setTimeout(checkValue, 100);

        function checkValue(){
            let input1 = wrapper.find('input').at(0);
            expect(input1.prop('value')).toEqual(145);
        }


    });


});
