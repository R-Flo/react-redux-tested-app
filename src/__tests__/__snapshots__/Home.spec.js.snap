// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOME snapshot It renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Home
    output={10}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Using React and Redux
        </h2>,
        <div>
          Input 1:
          <input
            placeholder="Input 1"
            type="text"
          />
        </div>,
        <div>
          Input 2 :
          <input
            placeholder="Input 2"
            type="text"
          />
        </div>,
        <div>
          Output :
          <input
            placeholder="Output"
            readOnly={true}
            type="text"
            value={10}
          />
        </div>,
        <div>
          <button
            id="add"
            onClick={[Function]}
          >
            Add
          </button>
          <button
            id="subtract"
            onClick={[Function]}
          >
            Subtract
          </button>
          <button
            id="fetchNumber"
            onClick={[Function]}
          >
            FETCH INPUT 1 NUMBER
          </button>
        </div>,
        <hr />,
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Using React and Redux",
        },
        "ref": null,
        "rendered": "Using React and Redux",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Input 1:",
            <input
              placeholder="Input 1"
              type="text"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Input 1:",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "placeholder": "Input 1",
              "type": "text",
            },
            "ref": "input1",
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Input 2 :",
            <input
              placeholder="Input 2"
              type="text"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Input 2 :",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "placeholder": "Input 2",
              "type": "text",
            },
            "ref": "input2",
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Output :",
            <input
              placeholder="Output"
              readOnly={true}
              type="text"
              value={10}
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Output :",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "placeholder": "Output",
              "readOnly": true,
              "type": "text",
              "value": 10,
            },
            "ref": "output",
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              id="add"
              onClick={[Function]}
            >
              Add
            </button>,
            <button
              id="subtract"
              onClick={[Function]}
            >
              Subtract
            </button>,
            <button
              id="fetchNumber"
              onClick={[Function]}
            >
              FETCH INPUT 1 NUMBER
            </button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add",
              "id": "add",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Subtract",
              "id": "subtract",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Subtract",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "FETCH INPUT 1 NUMBER",
              "id": "fetchNumber",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "FETCH INPUT 1 NUMBER",
            "type": "button",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "hr",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Using React and Redux
          </h2>,
          <div>
            Input 1:
            <input
              placeholder="Input 1"
              type="text"
            />
          </div>,
          <div>
            Input 2 :
            <input
              placeholder="Input 2"
              type="text"
            />
          </div>,
          <div>
            Output :
            <input
              placeholder="Output"
              readOnly={true}
              type="text"
              value={10}
            />
          </div>,
          <div>
            <button
              id="add"
              onClick={[Function]}
            >
              Add
            </button>
            <button
              id="subtract"
              onClick={[Function]}
            >
              Subtract
            </button>
            <button
              id="fetchNumber"
              onClick={[Function]}
            >
              FETCH INPUT 1 NUMBER
            </button>
          </div>,
          <hr />,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Using React and Redux",
          },
          "ref": null,
          "rendered": "Using React and Redux",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Input 1:",
              <input
                placeholder="Input 1"
                type="text"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Input 1:",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "placeholder": "Input 1",
                "type": "text",
              },
              "ref": "input1",
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Input 2 :",
              <input
                placeholder="Input 2"
                type="text"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Input 2 :",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "placeholder": "Input 2",
                "type": "text",
              },
              "ref": "input2",
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Output :",
              <input
                placeholder="Output"
                readOnly={true}
                type="text"
                value={10}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Output :",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "placeholder": "Output",
                "readOnly": true,
                "type": "text",
                "value": 10,
              },
              "ref": "output",
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                id="add"
                onClick={[Function]}
              >
                Add
              </button>,
              <button
                id="subtract"
                onClick={[Function]}
              >
                Subtract
              </button>,
              <button
                id="fetchNumber"
                onClick={[Function]}
              >
                FETCH INPUT 1 NUMBER
              </button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add",
                "id": "add",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Subtract",
                "id": "subtract",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Subtract",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "FETCH INPUT 1 NUMBER",
                "id": "fetchNumber",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "FETCH INPUT 1 NUMBER",
              "type": "button",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "hr",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
